### PR TITLE
Skeleton of Element Query Language

### PR DESCRIPTION
**Pull request by stephanenicolas**
First bricks of the Element Query Language, trying to structure it so that it can scale:
* an entity to query an element given a path
* an entity to get a path of an element
* unit tests for both 
* simplified unit tests via a few util classes
* started to document.

The element query language will be used to persist the state of annotation processors between builds: it is a way to retrieve annotations that we already processed in previous builds.

---
I am thinking more and more that it would be possible to use indices instead of names, but it doesn't change much this PR which is more about the structure of the module.

Indices instead of names: as we are describing things that didn't change, my guess is that javac will be returning the same annotation processing result for the same source file. Hence, we should be able to use indices, which would probably speed up things a lot (as it would avoid loops during retrieval).

### CODE REVIEW

**Comment by wangwh03** [comment]
Two spaces for indentation instead of four in Java files?

File: element-query-language/build.gradle [hunk 1]
```diff
@@ -0,0 +1,22 @@
+apply plugin: 'java'
```
Comment (stephanenicolas) [inline element-query-language/build.gradle]: Kotlin you mean ? Well, I think here we have to use java to make it really easy to integrate. More over as it's a java tool...
;)

**Comment by stephanenicolas** [comment]
@wangwh03 good point. We gotta setup a code formatting for this code. I will ask gradle which one they use. Otherwise, I really favor the google formatter. I can get that done in a different PR. 
I will open an issue for the formatting and quality tools, do not take this into account to review the PR.

File: element-query-language/build.gradle (hunk 1 above)
Comment (gstevey) [inline element-query-language/build.gradle]: As much as I like Kotlin (especially for the tests), I think it's probably best to use Java everywhere.

File: element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
+package com.gradle.incrap;
```
Comment (gstevey) [inline element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java]: name change -> incap

Also it's org.gradle (in the Gradle sources).

File: element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+import static com.gradle.incrap.QueryLanguage.LOCATOR_ANNOTATION_ON_FIELD;
+import static javax.lang.model.util.ElementFilter.fieldsIn;
+
+/**
+ * Retrieves annotation using the
+ * element query language.
+ */
+public class AnnotationFinder {
+  private static final String TAG = AnnotationFinder.class.getName();
```
Comment (gstevey) [inline element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java]: To be consistent with Gradle and javac itself, we should use 4-space indents.

File: element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+   * @return the annotation that matches this query or null if not found.
+   */
+  private TypeElement lookupAnnotationOnField(String subQuery) {
+    log("Looking up annotation on field %s", subQuery);
+    String[] split = subQuery.split(":");
+    String className = split[0];
+    String fieldName = split[1];
+    String annotationTypeName = split[2];
+
+    //find type
```
Comment (gstevey) [inline element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java]: Space after // in all comments -- the Google standard, and also used in all of the JDK sources.

Gradle does it wrong. :-)

File: element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    //find member in type
+    List<? extends Element> allMembers = elementUtils.getAllMembers(typeElement);
+    List<VariableElement> fields = fieldsIn(allMembers);
+    Element fieldElement = fieldElementByName(fields, fieldName);
+    if (fieldElement == null) {
+      return null;
+    }
+
+    //find annotation on member
+    final List<? extends AnnotationMirror> annotationMirrors = fieldElement.getAnnotationMirrors();
```
Comment (gstevey) [inline element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java]: I think marking locals as final has fallen out of fashion, except when required for closure capture.  I'd remove anything extraneous.

File: element-query-language/src/main/java/com/gradle/incrap/LogWrapper.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
+package com.gradle.incrap;
+
+//TODO: it's been a long time I didn't use a log framework, which one to take ?
```
Comment (gstevey) [inline element-query-language/src/main/java/com/gradle/incrap/LogWrapper.java]: We should avoid dependencies, so we have to use java.util.logging.  I'll set it up later.

File: element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java (hunk 3 above)
Comment (stephanenicolas) [inline element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java]: If it's the only constraint, we could use the google formatter, it has a single option for 2 or 4 spaces (as android uses 4).

File: element-query-language/src/main/java/com/gradle/incrap/LogWrapper.java (hunk 6 above)
Comment (stephanenicolas) [inline element-query-language/src/main/java/com/gradle/incrap/LogWrapper.java]: So it means that we can merge and fix that later @steveyegge ?

**Comment by stephanenicolas** [comment]
After all comments are adressed, do you guys still wanna review or I can merge ? 
We should try to be explicit about this in our reviews to speed up things.

**Comment by gstevey** [comment]
I'm @gstevey now, and yeah, normally I'd check the box saying "merge is OK", but I wanted to make absolutely sure the name change happens.  Usually it's OK to just submit after addressing review comments.

File: element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java (hunk 2 above)
Comment (stephanenicolas) [inline element-query-language/src/main/java/com/gradle/incrap/AnnotationFinder.java]: done

**Comment by stephanenicolas** [comment]
@gstevey @wangwh03 : Is this PR ok for you ?
I have applied the google formatter using 4 spaces (see PR: #8) and renamed to incap.

File: element-query-language/build.gradle (hunk 1 above)
Comment (wangwh03) [inline element-query-language/build.gradle]: Sorry was just thinking of using the following syntax for build.gradle like in Cedric's blog:
```
tasks.withType(JavaCompile) {
   options.incremental = true // one flag, and things will get MUCH faster
}
```
I am OK with go with what we have though.

File: element-query-language/src/main/java/com/gradle/incap/AnnotationPathEncoder.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import javax.lang.model.element.Element;
+
+public class AnnotationPathEncoder {
+
+    public String encodeAnnotationOnField(
+            Element fieldElement, Class<? extends Annotation> annotationClass) {
+        final String className = fieldElement.getEnclosingElement().toString();
+        final String fieldName = fieldElement.getSimpleName().toString();
+        final String annotationTypeName = annotationClass.getName();
+        return new StringBuilder() //
```
Comment (wangwh03) [inline element-query-language/src/main/java/com/gradle/incap/AnnotationPathEncoder.java]: We can probably remove // at the end of this line :)

**Comment by wangwh03** [comment]
Looks good to merge from my side. :shipit:

**Comment by gstevey** [comment]
Go ahead and merge when you're ready on this one.

**Pull request merged**